swap/src/bin/swap.rs

```diff
@@ -394,74 +409,144 @@ mod tests {
+
+        assert_eq!((amount, fees), (expected_amount, expected_fees))
     }
 
     #[tokio::test]
     async fn given_initial_balance_above_max_quantity_swaps_max_quantity() {
         let _guard = subscriber::set_default(tracing_subscriber::fmt().with_test_writer().finish());
 
-        let amount = determine_btc_to_swap(
-            async { Ok(quote_with_max(0.01)) },
-            async { Ok(Amount::from_btc(0.1)?) },
+        let (amount, fees) = determine_btc_to_swap(
+            quote_with_max(0.01),
+            async { Ok(Amount::from_btc(0.1001)?) },
             async { panic!("should not request new address when initial balance is > 0") },
             async { panic!("should not wait for deposit when initial balance > 0") },
-            async { Ok(Amount::from_btc(0.09)?) },
+            Amount::from_btc(0.1).unwrap(),
         )
         .await
         .unwrap();
 
-        assert_eq!(amount, Amount::from_btc(0.01).unwrap())
+        let expected_amount = Amount::from_btc(0.01).unwrap();
+        let expected_fees = Amount::from_btc(0.0001).unwrap();
+
+        assert_eq!((amount, fees), (expected_amount, expected_fees))
+    }
+
+    #[tokio::test]
+    async fn given_no_initial_balance_then_min_wait_for_sufficient_deposit() {
+        let _guard = subscriber::set_default(tracing_subscriber::fmt().with_test_writer().finish());
+
+        let (amount, fees) = determine_btc_to_swap(
+            quote_with_min(0.01),
+            async { Ok(Amount::from_btc(0.0101)?) },
+            get_dummy_address(),
+            async { Ok(Amount::from_btc(0.01)?) },
+            Amount::ZERO,
+        )
+        .await
+        .unwrap();
+
+        let expected_amount = Amount::from_btc(0.01).unwrap();
+        let expected_fees = Amount::from_btc(0.0001).unwrap();
+
+        assert_eq!((amount, fees), (expected_amount, expected_fees))
+    }
+
+    #[tokio::test]
+    async fn given_balance_less_then_min_wait_for_sufficient_deposit() {
+        let _guard = subscriber::set_default(tracing_subscriber::fmt().with_test_writer().finish());
+
+        let (amount, fees) = determine_btc_to_swap(
+            quote_with_min(0.01),
+            async { Ok(Amount::from_btc(0.0101)?) },
+            get_dummy_address(),
+            async { Ok(Amount::from_btc(0.01)?) },
+            Amount::from_btc(0.0001).unwrap(),
+        )
+        .await
+        .unwrap();
+
+        let expected_amount = Amount::from_btc(0.01).unwrap();
+        let expected_fees = Amount::from_btc(0.0001).unwrap();
+
+        assert_eq!((amount, fees), (expected_amount, expected_fees))
+    }
+
+    #[tokio::test]
+    async fn given_no_balance_and_transfers_less_than_min_keep_waiting_till_enough() {
+        // TODO: We currently cannot depict this test because logic was moved
+        // into waiting for the deposit, but that logic cannot be move into
+        // determine_btc_to_swap because we run into lifetime issues with
```

Comment:
A future can only be used once, if you need it multiple times, you need to add a layer of indirection that gives you a new future every time, aka a lambda.